File: Project_Euler/solutions/301_400/problem351/solution.py
"""

この解法はO(NlogN)であるため、実行速度の観点から以下のPyの実装ではH(1e6)を求めるにとどまっています。
MAXN = 1e8とすることで所望の回答を得ることができますが、Pythonでは1分以上時間がかかることが予想されます。

"""
MAXN = int(1e6)
ISPRIME = [True]*(MAXN + 1)
PHI = [i for i in range(MAXN + 1)]


ISPRIME[0] = ISPRIME[1] = False

for p in range(2, MAXN + 1):
    if ISPRIME[p] == True:
        PHI[p] = p - 1
        for j in range(2*p, MAXN + 1, p):
            ISPRIME[j] = False
            PHI[j] //= p
            PHI[j] *= (p - 1)


M = MAXN
ans = M*M
for i in range(2, M + 1):
    ans -= 2*PHI[i]
ans -= 1
ans += M - 1
ans *= 3
print(ans)

File: Project_Euler/solutions/301_400/problem351/solution.cpp
/*

この解法はO(NlogN)です。なお、空間計算量はO(N)です。以下のC++実装ではおよそ30秒ほどで回答を得ることができます。

*/

#include <bits/stdc++.h>
#define int long long
#define rep(i,n) for(int i=0;i<(n);i++)
#define print(a) cout << (a) << endl
#define len(X) ((int)(X).size())
using namespace std;
const int MAXN = 1e8;

vector<bool> isprime(MAXN + 1,true);
vector<int> phi(MAXN + 1);

void setup(){
    isprime[0] = isprime[1] = false;
    rep(i,MAXN + 1){phi[i] = i;}
    for (int p = 2; p <= MAXN; p++){
        if (isprime[p] == true){
            phi[p] = p - 1;
            for (int j = 2*p;j <= MAXN;j+=p){isprime[j] = false;phi[j]/=p;phi[j]*=(p - 1);}
        }
    }
    return;
}

signed main(){
    int M = 1e8;
    setup();
    int ans = M*M;
    for (int j = 2; j <= M;j++){ans -= 2*phi[j];}
    ans--;
    ans += (M - 1);
    ans *= 3;
    print(ans);
    return 0;
}